React recipe

adriantoine opened the pull request:
Fixes #446

Hi!

Here is a first attempt to a React recipe. I'm not a native english speaker so, please carefully read the whole recipe and fix my grammar :stuck_out_tongue: : https://github.com/adriantoine/ava/blob/master/docs/recipes/react.md

I might have forgotten something, please let me know!

Also, as a native french speaker, I'll be happy to make the translation to French once this is merged.

thangngoc89 commented:
I think you should add a note about adding babel-preset-react to AVA in order to use JSX inside test files

adriantoine commented:
@thangngoc89 there is one in the first part "Setting up Babel": https://github.com/sindresorhus/ava/pull/747/files#diff-2cb79c7fb78b66228297358846395c3aR7

thangngoc89 commented:
@adriantoine sorry. I missed that

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+
+## Setting up Babel
+
+The first thing you need to do is to set up `babel` to transpile JSX code from the tests. To do that, I'd recommend the [babelrc recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/babelrc.md) using `babel-preset-react`. You can also have a look at this sample project configuration: https://github.com/adriantoine/ava-enzyme-demo
```
Linkify `babel-preset-react` and linkify `sample project configuration` instead of showing the full URL.

sindresorhus replied:
configuration => config

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+
+The first thing you need to do is to set up `babel` to transpile JSX code from the tests. To do that, I'd recommend the [babelrc recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/babelrc.md) using `babel-preset-react`. You can also have a look at this sample project configuration: https://github.com/adriantoine/ava-enzyme-demo
+
+## Enzyme
```
Needs to be more descriptive.

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+
+## Enzyme
+
+Let's see first how to implement how to use ava with one of the most popular React testing library: [enzyme](https://github.com/airbnb/enzyme).
```
`Let's see first` => `Let's first see`

ava => AVA

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,98 @@
+If you only plan to use `shallow` component rendering, you don't need extra setup. First install `enzyme`:
+
+```js
+npm i --save-dev enzyme react-addons-test-utils react-dom
````
`npm i` => `npm install`

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,98 @@
+```js
+import test from 'ava';
+import React from 'react';
+import { shallow } from 'enzyme';
````
```
import {shallow} from 'enzyme';
```

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import sinon from 'sinon';
+
```
remove empty line

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+      <div className="unique" />
+    </MyComponent>
+  );
+  t.truthy(wrapper.contains(<div className="unique" />);
```
Use `t.true` and missing a `)`

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+    <Foo onButtonClick={onButtonClick} />
+  );
+  wrapper.find('button').simulate('click');
+  t.truthy(onButtonClick.calledOnce);
```
`t.true`

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,98 @@
+});
+```
+
+`enzyme` also has a `mount` and `render` helper to test in an actual browser environment, if you want to use these helpers, you will have to setup a browser environment, to do so, you should check out the [browser testing recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/browser-testing.md).
````
`environment, to do so,` => `environment. To do so,`

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+
+`enzyme` also has a `mount` and `render` helper to test in an actual browser environment, if you want to use these helpers, you will have to setup a browser environment, to do so, you should check out the [browser testing recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/browser-testing.md).
+
+Here is a simple and minimal example of testing react components using `ava` and `enzyme` along with browser testing: https://github.com/adriantoine/ava-enzyme-demo
```
react => React

sindresorhus replied:
Don't put in plain links, linkify some text instead.

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+
+## Using JSX helpers
+
+There is another approach to testing React component, which is to use the [`react-element-to-jsx-string`](https://github.com/algolia/react-element-to-jsx-string) package to compare DOM trees as strings, like [`expect-jsx`](https://github.com/algolia/expect-jsx). [You can use `expect-jsx` with ava](https://github.com/sindresorhus/ava/issues/186#issuecomment-161317068) however it is nicer to use av ...
```
ava => AVA

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+  const actual = renderJSX(<MyComponent />);
+  const expected = JSX(
+    <div onClick={ noop } className='MyComponent'>
+      <span className='icon-star'/>
```
Inconsistent quote usage. You use double-quotes earlier.

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,98 @@
+});
+```
+
+Note that you have to use variables like `actual` and `expected` because `power-assert` doesn't handle JSX correctly.
````
Can you linkify `power-assert doesn't handle JSX correctly` to the relevant `power-assert` issue?

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+
+Let's see first how to implement how to use ava with one of the most popular React testing library: [enzyme](https://github.com/airbnb/enzyme).
+
+If you only plan to use `shallow` component rendering, you don't need extra setup. First install `enzyme`:
```
`need extra setup` => `need any extra setup`

sindresorhus replied:
`First install enzyme:` should be on a new line.

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,98 @@
+# Testing React components
+
+This recipe is giving guidelines to test React components.
```
This doesn't really have much value. It's just a longer description of the title. I would drop it or come up with something else.

sindresorhus commented:
Nice work @adriantoine. This looks very promising :)

// @MoOx @kentcdodds @istarkov

sindresorhus commented:
> Also, as a native french speaker, I'll be happy to make the translation to French once this is merged.

@forresst FYI ;)

adriantoine commented:
@sindresorhus thanks! I'll have a look at those comments

MoOx commented:
Nice.

MoOx commented:
I would add a note about how to play with webpack, loaders for assets etc by poiting users to this https://github.com/istarkov/babel-plugin-webpack-loaders#dynamic-config-path

MoOx commented:
I would also add a note on the second solution to let user know that this one can "simulate" events too ;)

kentcdodds commented:
> I would add a note about how to play with webpack

That's not really React specific. We should make that a recipe and reference it.

adriantoine commented:
@MoOx https://github.com/istarkov/babel-plugin-webpack-loaders#dynamic-config-path I think it should be added to the [babelrc](https://github.com/sindresorhus/ava/blob/master/docs/recipes/babelrc.md) recipe, rather than the react one.

adriantoine commented:
@MoOx ah yeah, I'll add an example with an event

adriantoine commented:
I fixed the issues raised by @sindresorhus by the way

forresst commented:
> Also, as a native french speaker, I'll be happy to make the translation to French once this is merged.
@adriantoine  It is with pleasure that I will review (And feel free to review my translations)

adriantoine commented:
@MoOx I've added an example with an event simulation: https://github.com/sindresorhus/ava/pull/747/commits/a729aeba805caceb588287d80408580b6a4a163d I've used `sinon` like in the Enzyme example to have some sort of equivalent, let me know what you think

MoOx commented:
Actually I prefer this simple method that involve nothing else than what we already have https://github.com/MoOx/jsx-test-helpers/blob/master/src/__tests__/index.js#L98

MoOx commented on docs/recipes/react.md:
```diff
@@ -90,6 +91,21 @@ test('renders three <Foo /> components', t => {
 });
+
+test("Can render & test a class handler on a child", t => {
+  const onButtonClick = sinon.spy();
```
about about t.plan() before and a simple t.pass as a button callback?

adriantoine replied:
Ah yeah makes sense, same for the Enzyme example, I will update

kentcdodds commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,113 @@
+
+## Using [Enzyme](https://github.com/airbnb/enzyme/)
+
+Let's first see how to use AVA with one of the most popular React testing library: [Enzyme](https://github.com/airbnb/enzyme).
```
> with one of the most popular React testing library

Should be:

> with one of the most popular React testing libraries

kentcdodds commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,113 @@
+
+First install [Enzyme required packages](https://github.com/airbnb/enzyme/#installation):
+
+```js
````
Does this need to be highlighted as `js`?

forresst replied:
console instead of js (is better)

kentcdodds commented:
I think it'd be useful to reference [my workshop](https://github.com/kentcdodds/react-ava-workshop) as well.

adriantoine commented:
So I removed `sinon` as @MoOx has suggested, fixed "library" => "libraries" and removed js formatting from the `npm install` command as @kentcdodds has suggested

@kentcdodds should I add a reference part with a list of links or should I add it to the intro?

kentcdodds commented:
Maybe you could say: "For an in depth guide of seeing up AVA with code coverage on a React project, see this workshop" or something like that. That'd probably go well at the start

MoOx commented:
Currently, I found that the example with onClick for both enzyme or jsx-test-helpers does not provide a relevant tests: instead of testing that a callback is called, we should test that render (because that's what React is about) is indeed different that for the same render without a onClick called.
And again, I also found that using `TestUtils.Simulate.click` is overkill according to the explanation I just gave.
For this reason, I think testing via calling props on the render is enough, as I am doing here https://github.com/MoOx/jsx-test-helpers/blob/ab9260513e36845c98aee897a5616926f4dbeebe/src/__tests__/index.js#L71-L110
It's very simple and much more focused on the end result, not what happen under the hood (React probably have well tested that click on elements that have `onClick` actually call the `onClick` callback).

petetnt commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+
+There is another approach to testing React component, which is to use the [`react-element-to-jsx-string`](https://github.com/algolia/react-element-to-jsx-string) package to compare DOM trees as strings, like [`expect-jsx`](https://github.com/algolia/expect-jsx). [You can use `expect-jsx` with AVA](https://github.com/sindresorhus/ava/issues/186#issuecomment-161317068) however it is nicer to use AV ...
+
+To do so you can use the [`jsx-test-helpers`](https://github.com/MoOx/jsx-test-helpers) library:
```
Might be worth it to mention that `jsx-test-helpers` has `react-addons-test-utils` as an dependency so you don't have to `npm install` it yourself. (Or to add it to the install clause regardless if you are going use the TestUtils)

kentcdodds commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+You can find an annotated test file [here](https://github.com/MoOx/jsx-test-helpers/blob/master/src/__tests__/index.js) with more examples.
+
+## Reference
+* [In depth guide of seeing up AVA with code coverage on a React project](https://github.com/kentcdodds/react-ava-workshop)
```
`seeing` should be `setting`

adriantoine commented:
@MoOx I agree that the examples are not great, they were roughly copied from several sources and I guess I should have made my own set of examples. I will update the document tonight.

However I think it's better to show very basic examples in this recipe and link to your test file or your repo's documentation for further usage because this is AVA's doc, not Enzyme's doc or `jsx-test-helpers` doc, we should just show how to integrate AVA with those tools, then you're free to update your `jsx-test-helpers` documentation to show further examples or use cases of your library.

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+import test from 'ava';
+import React from 'react';
+import TestUtils from 'react-addons-test-utils';
+import { noop, renderJSX, JSX } from 'jsx-test-helpers';
```
`import {noop, renderJSX, JSX} from 'jsx-test-helpers':`

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+import Foo from '../../Foo';
+
+test('Can render & test a class component', t => {
+  const actual = renderJSX(<MyComponent />);
```
`const actual = renderJSX(<MyComponent/>);` (notice removal of the space in closing tag).

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+  );
+  const expected = JSX(
+    <div className='Foo'>
+      <button onClick={ noop }>{'Click Me'}</button>
```
`<button onClick={noop}>{'Click Me'}</button>`

vdemedes replied:
I'm wondering, why `{'Click Me'}` is between `{'` and `'}`?

adriantoine replied:
That was copied from @MoOx example, it's a practice that is documented in [`eslint-plugin-react`'s `jsx/no-literals` rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-literals.md).
I'll change the example's anyway, as I said yesterday (didn't have time to do it yesterday evening)

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+test('Can render & test a class handler on a child', t => {
+  t.plan(1);
+  const actual = renderJSX(
+    <Foo onButtonClick={() => t.pass()} />,
```
`<Foo onButtonClick={() => t.pass()}/>`

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+import Foo from '../../Foo';
+
+test('renders three <Foo /> components', t => {
+  const wrapper = shallow(<MyComponent />);
```
`const wrapper = shallow(<MyComponent/>);`

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+});
+
+test('renders an `.icon-star`', t => {
+  const wrapper = shallow(<MyComponent />);
```
`const wrapper = shallow(<MyComponent/>);`

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+      <div className='unique' />
+    </MyComponent>
+  );
+  t.true(wrapper.contains(<div className='unique' />));
```
`t.true(wrapper.contains(<div className='unique'/>));`

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+  t.plan(1);
+  const wrapper = shallow(
+    <Foo onButtonClick={() => t.pass()} />
+  );
```
`const wrapper = shallow(<Foo onButtonClick={() => t.pass()}/>`

kevva commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,112 @@
+test('renders children when passed in', t => {
+  const wrapper = shallow(
+    <MyComponent>
+      <div className='unique' />
```
`<div className='unique'/>`

kevva commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,112 @@
+```
+
+and test your React components:
+```js
````
Add a linebreak between line `70` and `71`.

adriantoine commented:
@kevva @sindresorhus about spacing and formatting, is there any rules I should follow? Maybe the ones from [`eslint-config-xo-react`](https://github.com/sindresorhus/eslint-config-xo-react)?

kevva commented:
@adriantoine, yup. Preferably :).

adriantoine commented:
Ok so, I changed all examples and they are now very basic, the component I am testing is in the example as well, for clarity.

The reason why the examples are very basic is that, this is AVA's documentation, so it's good to show that those libraries exist and that they work well with AVA, and I'm showing how to set them up, but it shouldn't be about showing every single use of those external libraries, for that I added a link to their documentation.
From now on, if you're still not happy with one of the example, please provide one to replace it.

@kevva Both examples are valid on [`xo`](https://github.com/sindresorhus/xo) and [`xo-react`](https://github.com/sindresorhus/eslint-config-xo-react), except of the function called `JSX` in @MoOx's library, because `xo` doesn't allow functions to be capitalised.

Could you all review the recipe again?

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+
+## Setting up Babel
+
+The first thing you need to do is to set up `babel` to transpile JSX code from the tests. To do that, I'd recommend the [babelrc recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/babelrc.md) using [`babel-preset-react`](http://babeljs.io/docs/plugins/preset-react/). You can also have a look at this [sample project config](https://github.com/adriantoine/ava-enzyme-demo)
```
Maybe show an example of the config needed directly here so lazy people can just copy-paste something that works?

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+
+The first thing you need to do is to set up `babel` to transpile JSX code from the tests. To do that, I'd recommend the [babelrc recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/babelrc.md) using [`babel-preset-react`](http://babeljs.io/docs/plugins/preset-react/). You can also have a look at this [sample project config](https://github.com/adriantoine/ava-enzyme-demo)
+
+## Using [Enzyme](https://github.com/airbnb/enzyme/)
```
`https://github.com/airbnb/enzyme/` => `https://github.com/airbnb/enzyme`

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+
+Let's first see how to use AVA with one of the most popular React testing libraries: [Enzyme](https://github.com/airbnb/enzyme).
+
+If you only plan to use `shallow` component rendering, you don't need any extra setup.
```
Maybe linkify `shallow component rendering`? Not everyone might know what it means exactly.

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,124 @@
+First install [Enzyme required packages](https://github.com/airbnb/enzyme/#installation):
+
+```console
+npm install --save-dev enzyme react-addons-test-utils react-dom
````
```
$ npm install --save-dev enzyme react-addons-test-utils react-dom
```

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,124 @@
+npm install --save-dev enzyme react-addons-test-utils react-dom
+```
+
+and you can use Enzyme straight away:
````
And

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,124 @@
+});
+```
+
+Enzyme also has a `mount` and `render` helper to test in an actual browser environment, if you want to use these helpers, you will have to setup a browser environment. To do so, you should check out the [browser testing recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/browser-testing.md).
````
`, if you want` => `. If you want`

sindresorhus replied:
> Check out the browser testing recipe on how to do so.

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+
+Enzyme also has a `mount` and `render` helper to test in an actual browser environment, if you want to use these helpers, you will have to setup a browser environment. To do so, you should check out the [browser testing recipe](https://github.com/sindresorhus/ava/blob/master/docs/recipes/browser-testing.md).
+
+To see an example of AVA working together with Enzyme set up for browser testing, you can have a look at [this sample project](https://github.com/adriantoine/ava-enzyme-demo).
```
`with Enzyme` => `with Enzyme,`

sindresorhus replied:
`you can have a look at` => `have a look at`

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+
+To see an example of AVA working together with Enzyme set up for browser testing, you can have a look at [this sample project](https://github.com/adriantoine/ava-enzyme-demo).
+
+This is a basic example about how to integrate Enzyme with AVA, to get more information about using Enzyme for unit testing React component, please have a look at [Enzyme's documentation](http://airbnb.io/enzyme/).
```
`with AVA, to get more information about` => `with AVA. For more information about`

sindresorhus replied:
`please have a look at` => `have a look at the`

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,124 @@
+To do so you can use the [`jsx-test-helpers`](https://github.com/MoOx/jsx-test-helpers) library:
+
+```js
+npm install --save-dev jsx-test-helpers
````
```
$ npm install --save-dev jsx-test-helpers
```

sindresorhus commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,124 @@
+npm install --save-dev jsx-test-helpers
+```
+
+and test your React components:
````
And

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+
+Note that you have to use variables like `actual` and `expected` because [`power-assert` doesn't handle JSX correctly](https://github.com/power-assert-js/power-assert/issues/34).
+
+This is a basic example about how to use `jsx-test-helpers` with AVA, to see a more advanced usage of this library, please have a look at [this annotated test file](https://github.com/MoOx/jsx-test-helpers/blob/master/src/__tests__/index.js).
```
`with AVA, to see` => `with AVA. To see`

sindresorhus replied:
`please have a look` => `have a look`

sindresorhus commented on docs/recipes/react.md:
```diff
@@ -0,0 +1,124 @@
+This is a basic example about how to use `jsx-test-helpers` with AVA, to see a more advanced usage of this library, please have a look at [this annotated test file](https://github.com/MoOx/jsx-test-helpers/blob/master/src/__tests__/index.js).
+
+## Reference
+* [In depth guide of setting up AVA with code coverage on a React project](https://github.com/kentcdodds/react-ava-workshop)
```
`*` => `-`

forresst commented on docs/recipes/react.md:
````diff
@@ -0,0 +1,124 @@
+
+To do so you can use the [`jsx-test-helpers`](https://github.com/MoOx/jsx-test-helpers) library:
+
+```js
````
Replace `js` by `console`

adriantoine commented:
Thanks @sindresorhus and @forresst, [it's all fixed](https://github.com/adriantoine/ava/commit/ec58ef46f09b18cd508acacb4718c8f77d418f61)!

adriantoine commented:
Another thing is, an update has been made to [`unexpected-react`](https://github.com/bruderstein/unexpected-react) to make it work with React 15, I've created a repo to try it with AVA. It works well and I quite like the colourful JSX output:
![unexpected-react output](https://raw.githubusercontent.com/adriantoine/ava-unexpected-react-demo/master/.gh/output.png)
compared to [`jsx-test-helpers`](https://github.com/MoOx/jsx-test-helpers):
![unexpected-react output](https://raw.githubusercontent.com/adriantoine/ava-jsx-test-helpers-demo/master/.gh/output.png)

On the other hand, you have to install [`unexpected`](https://github.com/unexpectedjs/unexpected) and [`unexpected-react`](https://github.com/bruderstein/unexpected-react) to your project which are both really big compared to [`jsx-test-helpers`](https://github.com/MoOx/jsx-test-helpers), so it might not suit everyone.

Should I mention [`unexpected-react`](https://github.com/bruderstein/unexpected-react) just like [`expect-jsx`](https://github.com/algolia/expect-jsx) has been mentioned? Is it worth an extra section? (considering `expect-jsx` doesn't have its own section)

sindresorhus commented:
I think it's worth mentioning.

adriantoine commented:
I've updated the recipe with a new section for "Other assertion libraries" where I put `expect-jsx` and `unexpected-react`.

scarletsky commented:
Hello @adriantoine , I follow your react recipes with some problem.

It works as expected when your component is in the test file. But when you require the component from another file (some thing like `import Foo from './Foo'`), it will throw errors like https://github.com/sindresorhus/ava/issues/458.

Any guides ?

MoOx commented:
You might need to use `"babel": "inherit"`, see https://github.com/sindresorhus/ava#es2015-support

scarletsky commented:
@MoOx 
 Yes, `"babel": "inherit"` works when I add `"react"` to my .babelrc file.

But I want to know why my config do not work.
```js
// .babelrc
{
  "presets": ["es2015", "stage-1"],
  "plugins": [
    "transform-decorators-legacy"
  ],
  "env": {
    "development": {
      "presets": ["react-hmre"]
    }
  }
}

// package.json
{
  "ava": {
    "verbose": true,
    "files": [
      "tests/client/App.js"
    ],
    "require": [
      "babel-register",
      "babel-polyfill"
    ],
    "babel": {
      "babelrc": true,
      "presets": ["react"]
    }
  }
}
```

From the docs:

> When specifying the Babel config for your tests, you can set the babelrc option to true. This will merge the specified plugins with those from your babelrc.

According to my understanding, my config should work too.
Do I miss anything ?

MoOx commented:
Why don't you add react preset in your babelrc?

scarletsky commented:
@MoOx Because I use babel in server side. In client side, I do something like:
```js
// webpack.config.js
var babelrc = JSON.parse(
  fs.readFileSync('./.babelrc', 'utf8')
)
// .....
{
  test: /\.js$/,
  exclude: /node_modules/,
  loader: 'babel',
  query: Object.assign(
    {},
    babelrc,
    {
      presets: babelrc.presets.concat('react'),
      cacheDirectory: true
    }
  )
}
```

This works very well.

scarletsky commented:
@MoOx Can you tell me why my prev config do not work ?

MoOx commented:
Sorry I don't have the answer. I use a single babel config for browser and node (for tests I use babel-plugin-webpack-loaders)

sindresorhus commented:
Alright. Let's merge! Superb work on this one @adriantoine. I think a lot of people will benefit from the recipe :)

sindresorhus commented:
![leo-clapping](https://cloud.githubusercontent.com/assets/170270/14865404/b244b65c-0ce8-11e6-9910-eca8d3b5216a.gif)

adriantoine commented:
Great thanks! 😄

forresst commented:
Great @adriantoine 

> Also, as a native french speaker, I'll be happy to make the translation to French once this is merged.

@adriantoine It is with pleasure that I will review . Ping me !

adriantoine commented:
@forresst yep, I'm working on it!
